File: runtime/python/units/wasi/fd_allocate.py
# requires: wasi/errno_fs
def wasi_fd_allocate(self, fd, offset, length):
    io = self.fds.get(fd)
    if io is None or isinstance(io, self.WasiDir):
        return self.ERRNO_BADF
    # fallocate never shrinks: grow the file to offset+len when that exceeds the
    # current size, otherwise leave it untouched (ADR-40).
    try:
        if offset + length > os.fstat(io.fileno()).st_size:
            os.ftruncate(io.fileno(), offset + length)
    except OSError as e:
        return self.fs_errno(e)
    return self.ERRNO_SUCCESS

File: runtime/perl/units/wasi/fd_allocate.pl
# requires: wasi/errno_fs
sub wasi_fd_allocate {
    my ($self, $fd, $offset, $len) = @_;
    my $e = $self->{fds}{$fd};
    return ERRNO_BADF if !defined($e) || $e->{dir};
    # fallocate never shrinks: grow the file to offset+len when that
    # exceeds the current size, otherwise leave it untouched.
    my $size = (stat($e->{fh}))[7];
    return $self->fs_errno(0 + $!) unless defined $size;
    if ($offset + $len > $size) {
        truncate($e->{fh}, $offset + $len) or return $self->fs_errno(0 + $!);
    }
    return ERRNO_SUCCESS;
}
